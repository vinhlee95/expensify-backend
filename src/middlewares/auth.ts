import {NextFunction, Request, Response} from 'express'
import {getTokenFromRequest, verifyToken} from '../utils/auth'
import apiError from '../utils/apiError'
import UserModel from '../resources/user/user.model'

/**
 * Middleware to check user's token
 *
 * - User has valid permission
 *   - Find user from database and attach to req.user
 *
 * @param optional
 * 	- If set true, only check token if header or query contains token
 */
export const checkToken = (optional: boolean = false) => {
	return async (req: Request, res: Response, next: NextFunction) => {
		const token: string = getTokenFromRequest(req)

		if (optional && !token) {
			return next()
		}

		let payload: any

		try {
			payload = await verifyToken(token)

			const {id, tokenId} = payload

			const user = await UserModel.findOne({
				_id: id,
				'passport.tokenId': tokenId,
			})

			if (!user) {
				return next(
					apiError.unauthorized(
						'Cannot find user with this token. This token may have been blacklisted',
					),
				)
			}

			req.user = user

			return next()
		} catch (e) {
			return next(apiError.unauthorized('Invalid token'))
		}
	}
}
