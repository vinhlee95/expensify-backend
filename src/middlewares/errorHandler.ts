import {ErrorRequestHandler} from 'express'
import chalk from 'chalk'
import {ApiError} from '../utils/apiError'
import httpStatus from 'http-status'
import createLogger from '../utils/logger'
import {errorResponse} from '../utils/apiResponse'

const logger = createLogger(module)

/**
 * Middleware to parse Error
 *
 * @param err
 * @param req
 * @param res
 * @param next
 */
const parseError: ErrorRequestHandler = (err, req, res, next) => {
	if (!(err instanceof ApiError)) {
		const apiError = new ApiError(
			err.message || err.msg,
			err.status || httpStatus.INTERNAL_SERVER_ERROR,
		)
		return next(apiError)
	}

	return next(err)
}

/**
 * Middleware to send Error
 *
 * @param err
 * @param req
 * @param res
 * @param next
 */
const sendError: ErrorRequestHandler = (err, req, res, next) => {
	const {status, message} = err

	if (status === httpStatus.INTERNAL_SERVER_ERROR) {
		logger.error(err)
	} else {
		logger.debug(`${chalk.red('[ERROR]:')} ${message}`)
	}

	res.status(status).json(errorResponse(err))
}

export const errorHandler = [parseError, sendError]
