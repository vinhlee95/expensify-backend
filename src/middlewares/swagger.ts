import swaggerJSDoc from 'swagger-jsdoc'
import swaggerUi from 'swagger-ui-express'
import packageJson from '../../package.json'

/**
 * Config swagger jsdoc
 *
 */
const swaggerDefinition = {
	openapi: '3.0.1',
	info: {
		title: packageJson.name,
		version: packageJson.version,
		description: packageJson.description,
	},
	components: {
		securitySchemes: {
			jwt: {
				type: 'http',
				scheme: 'bearer',
				bearerFormat: 'JWT',
			},
		},
	},
	security: [{jwt: [] as string[]}],
}

const options = {
	swaggerDefinition,
	apis: ['./src/resources/**/*.ts', './src/docs/*.yaml'],
}

const swaggerSpec = swaggerJSDoc(options)

/**
 * Middleware to setup swagger docs
 *
 */
const swagger = [swaggerUi.serve, swaggerUi.setup(swaggerSpec)]

export default swagger
