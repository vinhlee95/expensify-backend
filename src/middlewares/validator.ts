import {RequestHandler} from 'express'
import {query, ErrorFormatter, validationResult} from 'express-validator/check'
import apiError from '../utils/apiError'

export enum Sort {
	asc = 'asc',
	desc = 'desc',
}

/**
 * Middleware to validate common queries
 */
export const validateCommonQueries = () => {
	return [
		query('sort', 'Invalid sort query')
			.optional()
			.isIn([Sort.asc, Sort.desc]),
		query('offset', 'Invalid pagination offset, offset must be a number')
			.optional()
			.toInt()
			.isInt({min: 0}),
		query(
			'limit',
			'Invalid pagination limit, limit must be a number and greater than 0',
		)
			.optional()
			.toInt()
			.isInt({min: 1}),
	]
}

/**
 * Middleware to return validation createError from express-validator
 *
 * @param req
 * @param res
 * @param next
 */
export const handleValidationError: RequestHandler = (req, res, next) => {
	const errors = validationResult(req)

	const errorFormatter: ErrorFormatter<string> = ({msg}) => {
		return msg
	}

	if (!errors.isEmpty()) {
		const errorMessage = errors.formatWith(errorFormatter).array()[0] as string

		return next(apiError.badRequest(errorMessage))
	}

	return next()
}
