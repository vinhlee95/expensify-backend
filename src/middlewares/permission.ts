import _ from 'lodash'
import {NextFunction, Request, RequestHandler, Response} from 'express'
import {checkToken} from './auth'
import {UserRole, UserStatus} from '../resources/user/user.interface'
import apiError, {ErrorCode} from '../utils/apiError'

/**
 * Declare user's permissions
 */
export enum Permission {
	UserRead = 'user:read',
	UserWrite = 'user:write',
}

type PermissionRole = {[key in UserRole]: Permission[]}

export const permissionRole: PermissionRole = {
	[UserRole.Admin]: [Permission.UserRead, Permission.UserWrite],
	[UserRole.User]: [Permission.UserRead],
}

/**
 * Middleware to check user's permissions
 *
 * @param permissions
 */
const checkPermission = (permissions?: [Permission]) => {
	return (req: Request, res: Response, next: NextFunction) => {
		const {user} = req

		if (!user) {
			return next(apiError.unauthorized())
		}

		const userPermissions = permissionRole[user.role as UserRole]
		const hasPermission =
			_.difference(permissions, userPermissions).length === 0

		if (hasPermission) {
			return next()
		} else {
			return next(
				apiError.forbidden(
					'User has no permission to perform this action',
					ErrorCode.notHasPermission,
				),
			)
		}
	}
}

/**
 * Middleware to check user's status
 *
 * @param permissions
 */
export const checkUserStatus: RequestHandler = (req, res, next) => {
	const {user} = req

	if (!user) {
		return next(apiError.unauthorized())
	}

	if (user.status === UserStatus.Active) {
		return next()
	} else {
		return next(
			apiError.forbidden(
				'User has not been activated',
				ErrorCode.notActiveUser,
			),
		)
	}
}

/**
 * Middleware to check user's token then permissions
 *
 * @param permissions
 */
export const protect = (permissions: [Permission]) => {
	return [checkToken(), checkUserStatus, checkPermission(permissions)]
}
