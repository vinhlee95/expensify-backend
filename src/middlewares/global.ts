import {Router} from 'express'
import {json, urlencoded} from 'body-parser'
import morgan from 'morgan'
import cors from 'cors'
import helmet from 'helmet'
import validator from 'express-validator'
import rateLimit from 'express-rate-limit'
import {morganStream} from '../utils/logger'
import config from '../config'
import {validateCommonQueries} from './validator'

const router = Router()

// Time and amount limit of requests to the API
const {timeLimit, amountLimit} = config.requestLimiter

/**
 * Global middlewares
 */
router.use(cors())
router.use(json())
router.use(urlencoded({extended: true}))
router.use(morgan('dev', {stream: morganStream}))
router.use(validator())
router.use(validateCommonQueries())
router.use(helmet())
router.use(
	rateLimit({
		windowMs: timeLimit,
		max: amountLimit,
	}),
)

export default router
