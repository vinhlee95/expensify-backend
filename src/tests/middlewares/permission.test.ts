import {permissionRole} from '../../middlewares/permission'

describe('[Permission]', () => {
	it('permission role', () => {
		expect(permissionRole).toMatchSnapshot()
	})
})
