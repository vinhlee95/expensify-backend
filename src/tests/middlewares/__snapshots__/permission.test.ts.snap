// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Permission] permission role 1`] = `
Object {
  "admin": Array [
    "user:read",
    "user:write",
  ],
  "user": Array [
    "user:read",
  ],
}
`;
